### PR TITLE
misc: moved to license-plan-based rate limits

### PR DESCRIPTION
**Pull request by sheensantoscapadngan**
# Description 📣
This PR will enable users in shared cloud to have different rate limits based on their organization's license plan

License Server PR:
https://github.com/Infisical/license-server/pull/21

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation

# Tests 🛠️

```sh
# Here's some code block to paste some code snippets
```

---

- [x] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

File: backend/src/server/plugins/inject-rate-limits.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      return;
+    }
+
+    const plan = await server.services.license.getPlan(req.auth.orgId);
+    const { rateLimits } = plan;
+
+    if (plan.customRateLimits) {
+      req.rateLimits = defaultRateLimiterConfig;
+      return;
+    }
```
Comment (sheensantoscapadngan) [inline backend/src/server/plugins/inject-rate-limits.ts:21]: this should only be the case when self-hosted/dedicated instance :D

File: backend/src/ee/services/license/licence-fns.ts [hunk 2]
```diff
@@ -40,7 +40,17 @@ export const getDefaultOnPremFeatures = (): TFeatureSet => ({
   secretRotation: true,
   caCrl: false,
   instanceUserManagement: false,
-  externalKms: false
+  externalKms: false,
+  rateLimits: {
+    readLimit: 60,
+    publicEndpointLimit: 30,
+    writeLimit: 200,
```
Comment (maidul98) [inline backend/src/ee/services/license/licence-fns.ts:47]: what is the difference between write vs creation end points?

**Review by maidul98 — COMMENTED** [review]
Looks go to me. Although it looks like the whole org has a single rate limit now. Ideally it is per identity (user or machine) because right now let's say my org can have TPS of 10 for write but now every machine is subject to the same rate limit. That would mean that if i have machine A and machine B apart of the same org, their requests cannot go past 10 TPS individual

File: backend/src/server/config/rateLimiter.ts [hunk 3]
```diff
@@ -22,40 +21,46 @@ export const globalRateLimiterCfg = (): RateLimitPluginOptions => {
 // GET endpoints
 export const readLimit: RateLimitOptions = {
   timeWindow: 60 * 1000,
-  max: () => getRateLimiterConfig().readLimit,
+  hook: "preValidation",
```
[the remark this answers is not in the document]
Comment (sheensantoscapadngan) [reply backend/src/server/config/rateLimiter.ts:24]: exactly!

File: backend/src/ee/services/license/licence-fns.ts (hunk 2 above)
Comment (sheensantoscapadngan) [reply backend/src/ee/services/license/licence-fns.ts:47]: hmmm based on existing usage, creationLimit is for org, project, and machine identity creation while writeLimit is for everything else

**Review by maidul98 — APPROVED** [review]
lgtm

**Pull request merged**